(feat) Introduce Encounter Provider Control

kajambiya opened the pull request:
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
This PR introduces the Encounter Provider control.
When submitting an encounter, RFE by default picks the provider from the session(logged in user) however there are scenario where the provider is passed as part of the encounter. In this case, the default provider needs to be overridden. 

## Screenshots

https://github.com/openmrs/openmrs-form-engine-lib/assets/12844964/c48e5c54-1ed9-40c7-84f7-1afeb0ac3409



## Related Issue
<!-- Paste the link to the Jira ticket here if one exists. -->
<!-- https://issues.openmrs.org/browse/O3- -->

## Other
<!-- Anything not covered above -->

samuelmale commented on src/components/section/helpers.ts:
```diff
@@ -21,7 +21,9 @@ export function getFieldControlWithFallback(question: OHRIFormField) {
     question.questionOptions.rendering = 'encounter-location';
   }
-
+  if (question.type === 'encounterProvider') {
+    return getRegisteredControl('encounterProvider');
+  }
```
@kajambiya, there are two problematic issues with this:

1. We are gradually coupling the datatype property with the rendering property.
2. This function is performing more work than intended. Initially, it was meant to serve as a fallback to text rendering for fields lacking concepts. However, it now involves some form of mapping based on the datatype. Ideally, the `type` prop should configure the handler component, while the `rendering` should configure the actual control (the underlying component). Over time, this function may become bloated and challenging to maintain. Frankly, I don't see a good reason for this approach.

samuelmale replied:
I understand the issue, @kajambiya. There are a couple of tweaks that could potentially solve it.

Let's lay all the pieces on the table: Our approach to supporting derived controls (components based on data sources plus the ui-select-extended component) relies on a combination of several elements:
 
-  **The ui-select-extended component (USE)**
 This is a generic component used to flesh out our derived control
- **A Datasource**
This is responsible for loading arbitrary data ie. coded answers from the backend 
- **A Submission handler**
This is responsible for form field value binding and a factory of composite field values

### How do we put these together?
At the heart of the form-engine is the registry, which manages all components of the form-engine. 
To get a derived control or rendering we [introduced control templates](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/control-templates.ts#L3) that provide the default DS config, which are then used to [derive new controls](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/inbuiltControls.ts#L120-L124). 

This approach isn't perfect but seems to work perfectly and keeps us operating within the premises of the Form Registry. 

### The simpler solution (Feels a bit hacky)
We don't have to worry about the rendering and submission handler because, at the end of the day, the correct components will be injected based on the config you shared above. However, we have to figure out how to inject the datasource along with the [default config](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/control-templates.ts#L13-L25). Currently, if an instance of a USE component has no datasource config as part of the form field configuration, [we fallback to using the field rendering type](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/components/inputs/ui-select-extended/ui-select-extended.tsx#L49-L53) to identify an existing inbuilt DS template. The magic bullet is falling back to the `question.type` as opposed to the `question.questionOptions.rendering`. We just have to make sure that the control template name aligns with the value of the type prop. 
NB: This is a potential breaking change for rendering types such as "drug" and "problem" but easy to fix. 

### Second solution (I seem to prefer this one)
We can simply introduce a form schema adaptor or something along those lines. Its primary role could be to ensure that the schema is fine-tuned to what the renderer expects. Before rendering the form, we could perform some kind of schema validation and refinement.

Input:

```json
{
  "type": "encounterLocation",
  "label": "Facility name (site/satellite clinic required):",
  "id": "location",
  "required": "true",
  "questionOptions": {
    "rendering": "ui-select-extended"
  }
}
```

Output:

```json
{
  "type": "encounterLocation",
  "label": "Facility name (site/satellite clinic required):",
  "id": "location",
  "required": "true",
  "questionOptions": {
    "rendering": "encounter-location"
  }
}
```
Here we simply change the rendering to [encounter-location](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/inbuiltControls.ts#L60-L64) (Note that this rendering is not based on the USE) we are currently doing something similar [here](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/components/section/helpers.ts#L22-L23). Thoughts? cc: @ibacher @denniskigen

kajambiya replied:
Thanks @samuelmale 
I actually like both solutions. Let me wait to hear what other folks have to say

samuelmale commented on src/components/inputs/ui-select-extended/ui-select-extended.tsx:
```diff
@@ -45,7 +45,10 @@ const UISelectExtended: React.FC<OHRIFormFieldProps> = ({ question, handler, onC
   useEffect(() => {
-    const datasourceName = question.questionOptions?.datasource?.name;
+    let datasourceName = question.questionOptions?.datasource?.name;
+    if (question.type === 'encounterProvider') {
+      datasourceName = question.type;
+    }
```
@kajambiya Any good reason for making this component aware of the `encounterProvider` type?

kajambiya replied:
Same reason as highlighted above. Ideally if a control doesn't explicitly define a datasource, it will search for the datasource using the rendering (which will be the same name provided in the `controlTemplates`) but the AMPATH case makes it unique since they use type and not rendering to specify `encounterLocation`

samuelmale commented on src/api/api.ts:
```diff
@@ -71,6 +71,12 @@ export function getAllLocations(): Observable<{ uuid: string; display: string }[
   );
 }
 
+export function getEncounterProviders(): Observable<{ uuid: string; display: string }[]> {
+  return openmrsObservableFetch(`/ws/rest/v1/provider?v=custom:(uuid,display)`).pipe(
+    map(({ data }) => data['results']),
+  );
+}
```
Are you using this function?

kajambiya replied:
I had planned to use it my logic for getting the initial and previous value but that might change.

kajambiya replied:
Hi @samuelmale 
I've made some changes.
I haven't introduced a schema adaptor as you had earlier suggested. Realized I would have to first map through the entire schema(pages, sections, questions) to fine tune the schema and then I would have to map through the schema again to process sections. I therefore decided to add this transformation at the section level.
I've introduced the `transformSchema` function which checks every question in a section to see if it needs to be transformed. Let me know what you think.

eudson replied:
@samuelmale thank  you so much for these fruitful comments which I agree. I think there is a lot of things that need improvement on this code base and perhaps we should add some or most of this type of comments (in case they were not resolved yet) to the concept note (links to the PRs) so that we can work on them when we in Kampala next month for the hackathon. 

@kajambiya I remember that in out tech session today we agreed that this `transformSchema` belongs to the entry-point of the engine where the initial transformations happen. Unless you can't do there because of the additional looping your are talking about then that is fine otherwise it will slowdown the engine.

samuelmale replied:
>I remember that in out tech session today we agreed that this transformSchema belongs to the entry-point of the engine where the initial transformations happen.

I agree with you @eudson and the best place I can think of is [here](https://github.com/openmrs/openmrs-form-engine-lib/blob/main/src/hooks/useFormJson.tsx#L59)

kajambiya replied:
Done

samuelmale reviewed (commented):
Happy to see this taking shape, just a few early comments

denniskigen commented:
Very timely, thanks!

eudson commented on src/components/section/helpers.ts:
```diff
@@ -44,3 +37,19 @@ export function hasMissingConcept(question: OHRIFormField) {
+export function transformSchema(question: OHRIFormField) {
+  switch (question.type) {
+    case 'encounterProvider':
+      question.questionOptions.rendering = 'encounterProvider';
```
very minor: @kajambiya apologies for my OCD but why do we have the rendering types following different cases? camelCase and kebab-case?

eudson replied:
Also this function is not transforming the entire schema so I would suggest you rename it because it can be confusing and people can decide to add logic beyond just changing the field definitions

pirupius commented on src/api/types.ts:
```diff
@@ -199,6 +199,7 @@ export type RenderType =
   | 'group'
   | 'content-switcher'
   | 'encounter-location'
+  | 'encounterProvider'
```
```suggestion
  | 'encounter-provider'
```

pirupius commented on src/components/section/helpers.ts:
```diff
@@ -42,3 +37,19 @@ export function hasMissingConcept(question: OHRIFormField) {
+/**
+ * Make schema transformations especially for originally AFE schemas to match the RFE schema
+ */
+export function transformQuestion(question: OHRIFormField) {
```
🚀

pirupius commented on src/components/section/ohri-form-section.component.tsx:
```diff
@@ -111,7 +113,9 @@ const OHRIFormSection = ({ fields, onFieldChange }) => {
                           previousValue={previousFieldValue}
                           displayText={
-                            fieldDescriptor.questionOptions.rendering == 'checkbox'
+                            fieldDescriptor.questionOptions.rendering === 'date'
+                              ? formatDate(previousFieldValue)
+                              : fieldDescriptor.questionOptions.rendering == 'checkbox'
                               ? previousCheckboxDisplayValue
                               : previousFieldValue?.display
```
This is getting overly nested and looks like something that might grow with more unique rendering types

pirupius commented on src/datasources/provider-datasource.ts:
```diff
@@ -0,0 +1,27 @@
+
+export class ProviderDataSource extends BaseOpenMRSDataSource {
+  constructor() {
+    super('/ws/rest/v1/provider?v=custom:(uuid,display)');
```
```suggestion
    super(`${restBaseUrl}/provider?v=custom:(uuid,display)`);
```

kajambiya replied:
The file has only 1 single usage of the url

pirupius replied:
It's not about the usages. It's a pattern that has happened in all other monorepos to use `restBaseUrl` for the first potion of the rest api url.

kajambiya replied:
Done

samuelmale commented on src/components/section/helpers.ts:
```diff
@@ -42,3 +38,17 @@ export function hasMissingConcept(question: OHRIFormField) {
   );
 }
+
+function previousValueDisplayForCheckbox(previosValueItems: Object[]): String {
+  return previosValueItems.map((eachItem) => eachItem['display']).join(', ');
+}
+
+export const formatPreviousValueDisplayText = (question: OHRIFormField, previousFieldValue: any) => {
+  if (question.questionOptions.rendering === 'date') {
+    return formatDate(previousFieldValue);
+  }
+  if (question.questionOptions.rendering === 'checkbox') {
+    return Array.isArray(previousFieldValue) ? previousValueDisplayForCheckbox(previousFieldValue) : null;
+  }
+  return previousFieldValue?.display;
+};
```
```suggestion
export const formatFieldValueDisplay = (question: OHRIFormField, value: any) => {
  switch (question.questionOptions.rendering) {
    case 'date':
      return formatDate(value);
    case 'checkbox':
      return Array.isArray(value) ? value.map((entry) => entry['display']).join(', ') : value;
    default:
      return value?.display ?? value;
  }
};
```

kajambiya replied:
Done

samuelmale commented on src/datasources/provider-datasource.ts:
```diff
@@ -0,0 +1,27 @@
+    if (config?.tag) {
+      apiUrl = `${urlParts[0]}?tag=${config.tag}&${urlParts[1]}`;
+    }
+    //overwrite url if there's a uuid value, meaning we are in edit mode
+    if (uuid) {
+      apiUrl = `${urlParts[0]}/${uuid}?${urlParts[1]}`;
+    }
```
What are you trying to achieve here?

kajambiya replied:
This was a technical debt on the `ui-select-extended`
Remember we had the default one and the `searchable`.
There was an issue with the `searchable` one in edit mode - The saved value would't be picked because at the moment items would be empty since there's no attempt to search yet. However, the selected item was being picked as below:
`selectedItem={items.find((item) => item.uuid == field.value)}`
The solution was to add an optional parameter(uuid) to the `fetchData` function so that if the `ui-select-extended` is searchable, in edit mode, we pass the saved value(uuid) to the `fetchData` function so that we can pick the corresponding display value.

Not sure whether my explanation is not confusing

samuelmale replied:
Yes it makes sense to me, not a perfect way of solving this problem but it works.

kajambiya replied:
Agree.
I think we can brainstorm on the most perfect solution during the hackathon

samuelmale commented on src/hooks/useFormJson.tsx:
```diff
@@ -181,6 +184,38 @@ function updateFormJsonWithComponents(formJson: OHRIFormSchema, formComponents:
   });
 }
 
+const transformSchema = (formJson: OHRIFormSchema) => {
+  formJson.pages.forEach((page) => {
+    if (page.sections) {
+      let sections = page.sections;
+      sections.forEach((section) => {
+        if (section.questions) {
+          let questions = section.questions;
+          questions.map((question) => {
+            transformQuestion(question);
+          });
+        }
+      });
+    }
+  });
+};
```
This should be move to its own file and should be configurable. Transformers could implement some special interface:

```ts

interface FormSchemaTransformer {
   transform: (form: OHRIFormSchema) => OHRIFormSchema;
}

// AFE transformer

const AngularFormTransformer: FormSchemaTransformer = {
  transform: (form) => {
    // do transformations
    return form; 
  }
};

// transformer-form-schema.ts

function transformSchema(form: OHRIFormSchema) {
  const transformers = getRegisteredTransformers();
  transformers.forEach(transformer => {
    form = transformer.transform(form);
  });
  return form;
}
```

kajambiya replied:
Done

samuelmale commented on src/registry/registry.ts:
```diff
@@ -123,6 +142,48 @@ export async function getRegisteredFieldSubmissionHandler(type: string): Promise
+  const transformers = [];
+
+  // Fetch from cache if available
+  const cachedTransformers = registryCache.formTransformers;
+  if (Object.keys(cachedTransformers).length) {
+    for (const key in cachedTransformers) {
+      if (Object.prototype.hasOwnProperty.call(cachedTransformers, key)) {
+        transformers.push(cachedTransformers[key]);
+      }
+    }
+    return transformers;
+  }
+
+  //Get custom transformers
+  const formTransformersFromStore = getFormsStore().formTransformers;
+  if (Array.isArray(formTransformersFromStore)) {
+    const customTransformers = await Promise.all(
+      formTransformersFromStore.map(async (transformer) => {
+        const transformerImport = await transformer.load?.();
+        return transformerImport?.default;
+      }),
+    );
+    transformers.push(...customTransformers.filter((transformer) => transformer !== undefined));
+  }
+
+  if (inbuiltFormTransformers.length) {
+    for (const inbuiltTransformer of inbuiltFormTransformers) {
+      transformers.push(inbuiltTransformer.component);
+    }
+  }
+
+  // Cache all fetched transformers
+  for (const transformer of transformers) {
+    registryCache.formTransformers[
+      inbuiltFormTransformers.find((inbuiltTransformer) => inbuiltTransformer.component === transformer).name
+    ] = transformer;
+  }
```
I don't understand why this logic is this complicated. What makes transformers special from other registry components? As an example here is a similar implementation for field handlers:

```ts
export async function getRegisteredFieldSubmissionHandler(type: string): Promise<SubmissionHandler> {
  if (registryCache.fieldSubmissionHandlers[type]) {
    return registryCache.fieldSubmissionHandlers[type];
  }
  let handler = inbuiltFieldSubmissionHandlers.find((handler) => handler.type === type)?.component;
  // if undefined, try serching through the registered custom handlers
  if (!handler) {
    const handlerImport = await getFormsStore()
      .fieldSubmissionHandlers.find((handler) => handler.type === type)
      ?.load?.();
    handler = handlerImport?.default;
  }
  registryCache.fieldSubmissionHandlers[type] = handler;
  return handler;
}
```

kajambiya replied:
From your guidance earlier, we had to get all the transformers at once.
The difference with  `getRegisteredFieldSubmissionHandler` is that it only returns one handler at a time

samuelmale replied:
I could be wrong here but getting all transformers in one shot can be like "Hey registry store, give me all you got". Something like:

```ts

const cache = [];

async function getRegisteredTransformers() {
  if (cache.length) {
    return cache;
  }

  const customTransformers = await Promise.all(
    formTransformersFromStore.map(async (registration) => {
      try {
        const transformerImport = await registration.load?.();
        return transformerImport?.default;
      } catch (error) {
        console.error("Error loading custom transformer:", error);
        return null; // Handle loading error gracefully
      }
    })
  );
  const allTransformers = [...inbuiltFormTransformers.map((registration) => registration.component), ...customTransformers.filter(Boolean)]; // Filter out null values

  // Cache the transformers for future use
  cache.push(...allTransformers);

  return allTransformers;
}
```

samuelmale replied:
@kajambiya did you take action on the above?

samuelmale commented on src/api/types.ts:
```diff
@@ -332,3 +333,10 @@ export interface ProgramEnrollmentPayload {
   dateCompleted?: string;
   location: string;
 }
+
+/**
+ * Defines logic that transforms the form schema
+ */
+export interface FormSchemaTransformer {
+  transform: (form: OHRIFormSchema) => void;
+}
```
```suggestion

/**
 * A form schema transformer is used to bridge the gap caused by different variations of form schemas
 * in the OpenMRS JSON schema-based form-entry world. It fine-tunes custom schemas to be compliant
 * with the React Form Engine.
 */
export interface FormSchemaTransformer {
  /**
   * Transforms the raw schema to be compatible with the React Form Engine.
   */
  transform: (rawSchema: OHRIFormSchema) => OHRIFormSchema;
}
```

samuelmale commented on src/datasources/provider-datasource.ts:
```diff
@@ -0,0 +1,27 @@
+import { openmrsFetch, restBaseUrl } from '@openmrs/esm-framework';
+import { BaseOpenMRSDataSource } from './data-source';
+
+export class ProviderDataSource extends BaseOpenMRSDataSource {
+  constructor() {
+    super(`${restBaseUrl}/provider?v=custom:(uuid,display)`);
+  }
+
+  fetchData(searchTerm: string, config?: Record<string, any>, uuid?: string): Promise<any[]> {
+    let apiUrl = this.url;
+    const urlParts = apiUrl.split('?');
+    if (config?.tag) {
+      apiUrl = `${urlParts[0]}?tag=${config.tag}&${urlParts[1]}`;
+    }
+    //overwrite url if there's a uuid value, meaning we are in edit mode
+    if (uuid) {
+      apiUrl = `${urlParts[0]}/${uuid}?${urlParts[1]}`;
+    }
+
+    return openmrsFetch(searchTerm ? `${apiUrl}&q=${searchTerm}` : apiUrl).then(({ data }) => {
+      if (data.results) {
+        return data.results;
+      }
+      return data;
+    });
+  }
```
This implementation is unnecessarily complicated likely because of a poor abstraction in the `BaseOpenMRSDataSource`. I also see copy and paste residue eg. the part where you're applying a "tag"; providers don't support Tags. I would refactor to simply:

```ts
export class ProviderDataSource extends BaseOpenMRSDataSource {
  constructor() {
    super(null);
  }

  async fetchData(searchTerm: string, config?: Record<string, any>, uuid?: string): Promise<any[]> {
    const rep = 'v=custom:(uuid,display)';
    const url = uuid ? `${restBaseUrl}/provider/${uuid}?${rep}` : `${restBaseUrl}/provider?${rep}`;
    const { data } = await openmrsFetch(searchTerm ? `${url}&q=${searchTerm}` : url);
    if (data.results) {
      return data.results;
    }
    return Array.isArray(data) ? data : [data];
  }
}
```

samuelmale commented on src/registry/inbuilt-components/inbuiltTransformers.ts:
```diff
@@ -0,0 +1,10 @@
+import { RegistryItem } from '../registry';
+
+export const inbuiltFormTransformers: Array<RegistryItem<FormSchemaTransformer>> = [
+  {
+    name: 'AFESChemaTransformer',
+    component: AngularFormTransformer,
+  },
```
Nit:

```ts
 {
    name: 'AngularFormEngineSchemaTransformer',
    component: AngularFormEngineSchemaTransformer,
 },
```

samuelmale commented on src/transformers/AFETransformer.ts:
```diff
@@ -0,0 +1,35 @@
+import { FormSchemaTransformer, OHRIFormField, OHRIFormSchema } from '../api/types';
+
+export const AngularFormTransformer: FormSchemaTransformer = {
```
I would rename to something like: `AngularFormEngineSchemaTransformer` and also rename the file to: `angular-fe-schema-transformer.ts`

samuelmale commented on src/transformers/form-schema-transformer.ts:
```diff
@@ -0,0 +1,9 @@
+
+export async function transformSchema(form: OHRIFormSchema) {
+  const transformers = await getRegisteredformSchemaTransformers();
+  transformers.forEach((transformer) => {
+    transformer.transform(form);
+  });
```
I would use a reducer:

```ts
const refinedSchema = transformers.reduce((schema, transformer) => {
    return transformer.transform(schema);
  }, form);
  
return refinedSchema;
```

samuelmale reviewed (changes requested):
I'm starting to think we need to rethink the forms registry implementation. We should consider powering the registry by an IoC container like [inversify](https://github.com/inversify/InversifyJS). Otherwise thanks for the great work @kajambiya, just a few final comments.

kajambiya commented:
> I'm starting to believe that we need to rethink the forms registry implementation. We could consider powering the registry with an IoC container like [inversify](https://github.com/inversify/InversifyJS). Otherwise thanks for the great work @kajambiya, below are my final comments.

Hi @samuelmale 
I've addressed the issues save for the change requested on the `getRegisteredFieldSubmissionHandler` function for which I need more clarity. I've also included the logic for loading default values for `encounterProvider, encounterLocation & encounterDatetime.` Please review and revert.

samuelmale commented on src/submission-handlers/encounterProviderHandler.ts:
```diff
@@ -0,0 +1,34 @@
+    context: EncounterContext,
+  ) => {
+    if (encounter) {
+      const encounterProvider = encounter.encounterProviders[0]?.provider;
+      return encounterProvider ? encounterProvider.uuid : null;
```
```suggestion
      return encounter.encounterProviders[0]?.provider?.uuid;
      
```

samuelmale commented on src/submission-handlers/encounterProviderHandler.ts:
```diff
@@ -0,0 +1,34 @@
+  },
+
+  getDisplayValue: (field: OHRIFormField, value: any) => {
+    if (!field.value) {
+      return null;
+    }
+    return value;
```
```suggestion 
    return field.value;
```

samuelmale commented on src/hooks/useInitialValues.ts:
```diff
@@ -18,6 +18,7 @@ export function useInitialValues(
   const [isEncounterBindingComplete, setIsEncounterBindingComplete] = useState(
     encounterContext.sessionMode === 'enter',
   );
+  const sessionInferredValueTypes = ['encounterProvider', 'encounterDatetime', 'encounterLocation'];
```
```suggestion
  const encounterContextInitialisableTypes = ['encounterProvider', 'encounterDatetime', 'encounterLocation'];
```

samuelmale commented on src/registry/registry.ts:
```diff
@@ -83,6 +98,18 @@ export function registerExpressionHelper(name: string, fn: Function) {
   getFormsStore().expressionHelpers[name] = fn;
 }
 
+export function registereformSchemaTransformers(registration: FormScemaTransformerRegistration) {
```
```suggestion
export function registerformSchemaTransformer(registration: ComponentRegistration<FormSchemaTransformer>) {
```

samuelmale commented on src/registry/registry.ts:
```diff
@@ -83,6 +98,18 @@ export function registerExpressionHelper(name: string, fn: Function) {
 
+export function registereformSchemaTransformers(registration: FormScemaTransformerRegistration) {
+  const store = getFormsStore();
+  const existingIndex = store.formTransformers.findIndex((reg) => reg.name === registration.name);
+
+  if (existingIndex !== -1) {
+    // If registration with the same name exists, override it
+    store.formTransformers[existingIndex] = registration;
+  } else {
+    store.formTransformers.push(registration);
+  }
+}
```
This seems to imply that the transformers are stored in array yet they seem to be stored in a map. I would simply:

```ts
getFormsStore().formTransformers[registration.name] = registration;
```

kajambiya replied:
Hi @samuelmale 
If I got you right, it's actually an array.
This is how the FormsRegistryStoreState looks like:
```
  export interface FormsRegistryStoreState {
  controls: CustomControlRegistration[];
  fieldValidators: ComponentRegistration<FieldValidator>[];
  fieldSubmissionHandlers: FieldSubmissionHandlerRegistration[];
  postSubmissionActions: ComponentRegistration<PostSubmissionAction>[];
  dataSources: ComponentRegistration<DataSource<any>>[];
  expressionHelpers: Record<string, Function>;
  formSchemaTransformers: FormSchemaTransformerRegistration[];
}
```

I followed the same logic for submission handlers, the only difference being that for transformers we override
```
export function registerFieldSubmissionHandler(registration: FieldSubmissionHandlerRegistration) {
  getFormsStore().fieldSubmissionHandlers.push(registration);
}
```

samuelmale commented on src/registry/registry.ts:
```diff
@@ -32,13 +40,18 @@ export interface FieldSubmissionHandlerRegistration extends ComponentRegistratio
   type: string;
 }
 
+export interface FormScemaTransformerRegistration extends ComponentRegistration<FormSchemaTransformer> {
+  type: string;
+}
```
There are two issues:

1. There is a typo in the interface name
2. Do we need the `type` property?

eudson commented:
Thank you @samuelmale
